common/src/main/scala/uk/ac/wellcome/dynamo/VersionedDao.scala

```diff
@@ -2,57 +2,164 @@ package uk.ac.wellcome.dynamo
+import shapeless.{HList, LabelledGeneric, Poly1, Witness}
+import shapeless.labelled.FieldType
 import uk.ac.wellcome.models._
+import shapeless._
+import shapeless.ops.hlist.{LeftFolder, Mapper}
+import shapeless.ops.record.Remover
+import shapeless.ops.tuple.ToTraversable
+import shapeless.record._
+import uk.ac.wellcome.dynamo.UpdateExpressionGenerator.toUpdateExpressions
 import uk.ac.wellcome.models.aws.DynamoConfig
+import uk.ac.wellcome.type_classes.IdGetter.{createIdGetter, IdKey}
+import uk.ac.wellcome.type_classes.{IdGetter, VersionGetter, VersionUpdater}
 import uk.ac.wellcome.utils.GlobalExecutionContext.context
 
+import scala.collection.generic.CanBuildFrom
 import scala.concurrent.Future
 
+trait UpdateExpressionGenerator[T] {
+  def generateUpdateExpression(t: T): Option[UpdateExpression]
+}
+
+object UpdateExpressionGenerator {
+  val w = Witness(Symbol("id"))
+  type IdKey = w.T
+  type IdField = FieldType[IdKey, String]
+
+  object toUpdateExpressions extends Poly1 {
+    implicit def some[K <: Symbol, V](implicit witness: Witness.Aux[K],
+                                      dynamoFormat: DynamoFormat[V])
+      : Case.Aux[FieldType[K, V], UpdateExpression] = {
+
+      at[FieldType[K, V]] { fieldtype =>
+        {
+          val fieldValue: V = fieldtype
+          set(witness.value -> fieldValue)
+        }
+      }
+    }
+  }
+
+  object Folder extends Poly2 {
+    implicit def fold: Case.Aux[Option[UpdateExpression],
+                                UpdateExpression,
+                                Option[UpdateExpression]] =
+      at[Option[UpdateExpression], UpdateExpression] {
+        (maybeUpdateExpression, partUpdate) =>
+          maybeUpdateExpression match {
+            case Some(updateExpression) =>
+              Some(updateExpression and partUpdate)
+            case None => Some(partUpdate)
+          }
+      }
+  }
+
+  def apply[T](implicit generator: UpdateExpressionGenerator[T]) = generator
+
+  def create[T](f: (T) => Option[UpdateExpression]) =
+    new UpdateExpressionGenerator[T] {
+      override def generateUpdateExpression(record: T) = f(record)
+    }
+
+  implicit def elementUpdateExpressionGenerator[L <: HList,
+                                                A,
+                                                B <: HList,
+                                                C <: HList](
+    implicit remover: Remover.Aux[L, IdKey, (A, B)],
+    mapper: Mapper.Aux[toUpdateExpressions.type, B, C],
+    folder: LeftFolder.Aux[C,
+                           Option[UpdateExpression],
+                           Folder.type,
+                           Option[UpdateExpression]]) =
+    create { t: L =>
+      {
+        val recordAsHlist = t - 'id
+        val nonIdTaggedFields = recordAsHlist.map(toUpdateExpressions)
+
+        nonIdTaggedFields.foldLeft[Option[UpdateExpression]](None)(Folder)
+      }
+    }
+
+  implicit def productUpdateExpressionGenerator[C, T](
```

Comment:
Is "product" a case class, or something else?

Reply:
Yes, it's a case class in shapeless terminology